wpilibj/src/main/java/edu/wpi/first/wpilibj/sysid/SysIdRoutineLog.java

```diff
@@ -0,0 +1,140 @@
+   *     entry.
+   */
+  public SysIdRoutineLog(String logName) {
+    m_state = new StringLogEntry(DataLogManager.getLog(), "sysid-test-state-" + logName);
+    m_state.append(State.kNone.toString());
+  }
+
+  /** Possible state of a SysId routine. */
+  public enum State {
+    kQuasistaticForward("quasistatic-forward"),
+    kQuasistaticReverse("quasistatic-reverse"),
+    kDynamicForward("dynamic-forward"),
+    kDynamicReverse("dynamic-reverse"),
+    kNone("none");
+
+    private final String m_state;
+
+    State(String state) {
+      m_state = state;
+    }
+
+    @Override
+    public String toString() {
+      return m_state;
+    }
+  }
+
+  static class DataFrame {
+    public double m_voltage;
+    public double m_distance;
+    public double m_velocity;
+
+    DataFrame(double voltage, double distance, double velocity) {
+      m_voltage = voltage;
+      m_distance = distance;
+      m_velocity = velocity;
+    }
+
+    DataFrame() {
+      this(0, 0, 0);
+    }
+
+    static class DataFrameStruct implements Struct<DataFrame> {
+      @Override
+      public Class<DataFrame> getTypeClass() {
+        return DataFrame.class;
+      }
+
+      @Override
+      public String getTypeString() {
+        return "SysIdDataFrame";
+      }
+
+      @Override
+      public int getSize() {
+        return kSizeDouble * 3;
+      }
+
+      @Override
+      public String getSchema() {
+        return "double voltage; double distance; double velocity";
+      }
+
+      @Override
+      public DataFrame unpack(ByteBuffer bb) {
+        double voltage = bb.getDouble();
+        double distance = bb.getDouble();
+        double velocity = bb.getDouble();
+        return new DataFrame(voltage, distance, velocity);
+      }
+
+      @Override
+      public void pack(ByteBuffer bb, DataFrame value) {
+        bb.putDouble(value.m_voltage);
+        bb.putDouble(value.m_distance);
+        bb.putDouble(value.m_velocity);
+      }
+    }
+
+    static final DataFrameStruct struct = new DataFrameStruct();
+  }
+
+  @Override
+  public void recordFrame(
+      double voltage, double distance, double velocity, String motorName, String distanceUnit) {
+    var entry = m_logEntries.get(motorName);
+
+    if (entry == null) {
+      entry =
+          StructLogEntry.create(DataLogManager.getLog(), motorName, DataFrame.struct, distanceUnit);
+      m_logEntries.put(motorName, entry);
+    }
+
+    m_dataFrame.m_voltage = voltage;
+    m_dataFrame.m_distance = distance;
+    m_dataFrame.m_velocity = velocity;
+
+    entry.append(m_dataFrame);
+  }
+
+  /**
+   * Records the current state of the SysId test routine. Should be called once per iteration during
+   * tests with the type of the current test, and once upon test end with state `none`.
+   *
+   * @param state The current state of the SysId test routine.
+   */
+  public void recordState(State state) {
```

Comment:
I think this belongs in the MotorLog class, otherwise third-party log implementations can't log State information without re-implementing the sysid motions such as `quasistatic` and `dynamic`.

Reply:
Escape hatch has been integrated into `Config` to cover this, per discussion